### PR TITLE
mon/OSDMonitor: add flag `--yes-i-really-mean-it` for setting pool size 1

### PR DESCRIPTION
**Pull request by ideepika**
Adds option `mon_allow_pool_size_one` which will be disabled by default
to ensure pools are not configured without replicas.
If the user still wants to use pool size 1, they will have to change the
value of `mon_allow_pool_size_one` to true and then have to pass flag
`--yes-i-really-really-mean-it` to cli command:

Example:
`ceph osd pool test set size 1 --yes-i-really-really-mean-it`

Fixes: https://tracker.ceph.com/issues/44025
Signed-off-by: Deepika Upadhyay <dupadhya@redhat.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/options.cc [hunk 1]
```diff
@@ -1735,6 +1735,11 @@ std::vector<Option> get_global_options() {
     .add_see_also("osd_pool_default_size")
     .add_see_also("osd_pool_default_min_size"),
 
+    Option("mon_allow_pool_size_one", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
+    .set_default(true)//false)
```
Comment (neha-ojha) [inline src/common/options.cc]: I think we want to keep the default false

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -7925,10 +7925,22 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
...
+      return -EPERM;
+    }
     if (n <= 0 || n > 10) {
       ss << "pool size must be between 1 and 10";
       return -EINVAL;
     }
+    bool sure = false;
+    cmd_getval(cmdmap, "yes_i_really_really_mean_it", sure);
+    if (n == 1){
+    if (!sure){
```
Comment (neha-ojha) [inline src/mon/OSDMonitor.cc]: indent?

File: src/mon/MonCommands.h [hunk 3]
```diff
@@ -1030,7 +1030,8 @@ COMMAND("osd pool set "
 	"name=pool,type=CephPoolname "
 	"name=var,type=CephChoices,strings=size|min_size|pg_num|pgp_num|pgp_num_actual|crush_rule|hashpspool|nodelete|nopgchange|nosizechange|write_fadvise_dontneed|noscrub|nodeep-scrub|hit_set_type|hit_set_period|hit_set_count|hit_set_fpp|use_gmt_hitset|target_max_bytes|target_max_objects|cache_target_dirty_ratio|cache_target_dirty_high_ratio|cache_target_full_ratio|cache_min_flush_age|cache_min_evict_age|min_read_recency_for_promote|min_write_recency_for_promote|fast_read|hit_set_grade_decay_rate|hit_set_search_last_n|scrub_min_interval|scrub_max_interval|deep_scrub_interval|recovery_priority|recovery_op_priority|scrub_priority|compression_mode|compression_algorithm|compression_required_ratio|compression_max_blob_size|compression_min_blob_size|csum_type|csum_min_block|csum_max_block|allow_ec_overwrites|fingerprint_algorithm|pg_autoscale_mode|pg_autoscale_bias|pg_num_min|target_size_bytes|target_size_ratio "
 	"name=val,type=CephString "
-	"name=yes_i_really_mean_it,type=CephBool,req=false",
+	"name=yes_i_really_mean_it,type=CephBool,req=false "
+	"name=yes_i_really_really_mean_it,type=CephBool,req=false",
```
Comment (xiexingguo) [inline src/mon/MonCommands.h]: IMHO, I don't really like the syntax. 
FWIW,  can't you simply reuse the existing  yes_i_really_mean_it to achieve the same purpose?

File: src/common/options.cc (hunk 1 above)
Comment (ideepika) [reply src/common/options.cc]: ack.

File: src/mon/MonCommands.h (hunk 3 above)
Comment (ideepika) [reply src/mon/MonCommands.h]: @xiexingguo yeah, that would be good alternative, made necessary changes.

File: PendingReleaseNotes [hunk 4]
```diff
@@ -303,4 +303,9 @@
 
 * RGW S3: Support has been added for BlockPublicAccess set of APIs at a bucket
   level, currently blocking/ignoring public acls & policies are supported.
-  User/Account level apis are planned to be added in the future
\ No newline at end of file
+  User/Account level apis are planned to be added in the future
+
+* Monitor will now have a configuration option to allow setting pool size to 1
```
Comment (neha-ojha) [inline PendingReleaseNotes]:
```suggestion
* Monitors now have a configuration option ``mon_allow_pool_size_one`` to allow setting pool size to 1, which is disabled by default
```

File: PendingReleaseNotes [hunk 5]
```diff
@@ -303,4 +303,9 @@
...
+  which would be disabled by default. If enabled, the user will also have to pass
+  ``--yes-i-really-mean-it`` flag to ``osd pool set size`` if they want to
+  configure pool with no replicas.
```
Comment (neha-ojha) [inline PendingReleaseNotes]:
```suggestion
  configure a pool with no replicas.
```

File: src/mon/OSDMonitor.cc [hunk 6]
```diff
@@ -7925,10 +7925,22 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
...
+      return -EPERM;
+    }
     if (n <= 0 || n > 10) {
       ss << "pool size must be between 1 and 10";
       return -EINVAL;
     }
+    bool sure = false;
+    cmd_getval(cmdmap, "yes_i_really_mean_it", sure);
+    if (n == 1) {
+    if (!sure) {
```
Comment (neha-ojha) [inline src/mon/OSDMonitor.cc]: https://github.com/ceph/ceph/blob/db5b1d6c29a2ca4c3784397932adb4f8d24ba8ef/src/mon/OSDMonitor.cc#L7939 indentation still needs to be fixed

File: qa/standalone/mon/health-mute.sh [hunk 7]
```diff
@@ -38,7 +38,7 @@ function TEST_mute() {
     ceph -s
     ceph health | grep HEALTH_OK || return 1
     # test warning on setting pool size=1
-    ceph osd pool set foo size 1
+    ceph osd pool set foo size 1 --yes-i-really-mean-it
```
Comment (xiexingguo) [inline qa/standalone/mon/health-mute.sh]: I think you should instead set "mon_allow_pool_size_one = true" somewhere to let these kind of tests pass?

Comment (xiexingguo) [reply qa/standalone/mon/health-mute.sh]: I think you should also set "mon_allow_pool_size_one = true" somewhere to let these kind of tests pass?

Comment (xiexingguo) [inline qa/standalone/mon/health-mute.sh]: I think you should also set "mon_allow_pool_size_one = true" somewhere to let these kind of tests pass?

File: PendingReleaseNotes [hunk 8]
```diff
@@ -303,4 +303,9 @@
 
 * RGW S3: Support has been added for BlockPublicAccess set of APIs at a bucket
   level, currently blocking/ignoring public acls & policies are supported.
-  User/Account level apis are planned to be added in the future
\ No newline at end of file
+  User/Account level apis are planned to be added in the future
+
+* Monitors now have a configuration option ``mon_allow_pool_size_one`` to allow setting pool size to 1, which is disabled by default
+  which would be disabled by default. If enabled, the user will also have to pass
```
Comment (gregsfortytwo) [inline PendingReleaseNotes]: This sentence has two endings?

File: src/mon/OSDMonitor.cc [hunk 9]
```diff
@@ -7925,10 +7925,22 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
       ss << "error parsing integer value '" << val << "': " << interr;
       return -EINVAL;
     }
+    if (!g_conf()->mon_allow_pool_size_one) {
+      ss << "configuring pool without replicas is disabled by default; you must first set the mon_allow_pool_size_one config option to true before you can set pool size to 1";
```
Comment (gregsfortytwo) [inline src/mon/OSDMonitor.cc]: I actually think we don't want to reference the config option here. There's just no point when setting pool size 1 is a good idea and it's okay to make it hard for users to do dumb things. Tell them it's disabled by default and leave it at that.

Comment (gregsfortytwo) [reply src/mon/OSDMonitor.cc]: Also this text block output isn't covered by the check for "n == 1" so it'll show up whenever anybody tries to set any pool size; it needs to get moved?

Comment (ideepika) [reply src/mon/OSDMonitor.cc]: yep, missed that check! thanks!

File: qa/standalone/mon/health-mute.sh (hunk 7 above)
Comment (ideepika) [reply qa/standalone/mon/health-mute.sh]: yep, thanks! updated `mon_allow_osd_size_one = true` for vstart.

**Comment by gregsfortytwo** [comment]
> This change doesn't fit well with the operator model and its declarative nature but I suspect that this change is somehow desired and good for users.

Pools with no replicas is just never a good idea. It's technically possible to run but any OSD down blocks IO on its portion of data and actually losing one means both data loss and horrifying manual gyrations to try and get back what we can. Even in a cloud environment with eg EBS backing disks I can't think of any way this would work.

**Comment by leseb** [comment]
> > This change doesn't fit well with the operator model and its declarative nature but I suspect that this change is somehow desired and good for users.
> 
> Pools with no replicas is just never a good idea. It's technically possible to run but any OSD down blocks IO on its portion of data and actually losing one means both data loss and horrifying manual gyrations to try and get back what we can. Even in a cloud environment with eg EBS backing disks I can't think of any way this would work.

I couldn't agree more! Again I'm not saying it's a bad thing it's just that Rook for instead would have to suppress that if a user asks for pool size of 1 and we will only rely on the health warning.

**Comment by ideepika** [comment]
Rerun: 
Test Run: ideepika-2020-02-20_07:04:09-rados-wip-deepika-osd-pool-set-config-2-distro-basic-smithi
=================================================================
logs:    http://qa-proxy.ceph.com/teuthology/ideepika-2020-02-20_07:04:09-rados-wip-deepika-osd-pool-set-config-2-distro-basic-smithi/
passed:  4
=================================================================
Test Run: ideepika-2020-02-17_14:13:20-rados-wip-deepika-testing-17-02-2020-distro-basic-smithi
=================================================================
logs:    http://qa-proxy.ceph.com/teuthology/ideepika-2020-02-17_14:13:20-rados-wip-deepika-testing-17-02-2020-distro-basic-smithi/
failed:  7
passed:  52
=================================================================

File: src/common/legacy_config_opts.h [hunk 10]
```diff
@@ -236,6 +236,7 @@ OPTION(mon_osd_backfillfull_ratio, OPT_FLOAT) // what % full makes an OSD backfi
 OPTION(mon_osd_nearfull_ratio, OPT_FLOAT) // what % full makes an OSD near full
 OPTION(mon_osd_initial_require_min_compat_client, OPT_STR)
 OPTION(mon_allow_pool_delete, OPT_BOOL) // allow pool deletion
+OPTION(mon_allow_pool_size_one, OPT_BOOL) // allow setting pool size as 1
```
Comment (tchaikov) [inline src/common/legacy_config_opts.h]: no need to add to legacy options.

File: src/mon/OSDMonitor.cc [hunk 11]
```diff
@@ -7929,6 +7929,18 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
       ss << "pool size must be between 1 and 10";
       return -EINVAL;
     }
+    if (n == 1) {
+      bool sure = false;
+      cmd_getval(cmdmap, "yes_i_really_mean_it", sure);
+      if (!g_conf()->mon_allow_pool_size_one) {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: use the new-style `get_val<>` for retrieving the options in non-critical paths.

File: src/mon/OSDMonitor.cc [hunk 12]
```diff
@@ -7929,6 +7929,18 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
       ss << "pool size must be between 1 and 10";
       return -EINVAL;
     }
+    if (n == 1) {
+      bool sure = false;
+      cmd_getval(cmdmap, "yes_i_really_mean_it", sure);
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: please move the definition close to where this variable is used.

File: src/mon/OSDMonitor.cc [hunk 13]
```diff
@@ -7929,6 +7929,18 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
...
+      if (!g_conf()->mon_allow_pool_size_one) {
+	ss << "configuring pool size as 1 is disabled by default.";
+	return -EPERM;
+      }
+      if (!sure) {
+	ss << "WARNING: setting pool size 1 could lead to data loss without recovery. If you are *ABSOLUTELY CERTAIN* that is what you want, pass the flag --yes-i-really-really-mean-it.";
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: wrap at char 80.

File: qa/tasks/ceph.conf.template [hunk 14]
```diff
@@ -88,6 +89,7 @@
 	mon osd reporter subtree level = osd
 	mon osd prime pg temp = true
 	mon reweight min bytes per osd = 10
+	mon_allow_pool_size_one = true
```
Comment (tchaikov) [inline qa/tasks/ceph.conf.template]: this is not necessary.

Comment (ideepika) [reply qa/tasks/ceph.conf.template]: aah, sorry missed, thanks! ◕‿◕

File: PendingReleaseNotes [hunk 15]
```diff
@@ -306,4 +306,9 @@
 
 * RGW S3: Support has been added for BlockPublicAccess set of APIs at a bucket
   level, currently blocking/ignoring public acls & policies are supported.
-  User/Account level apis are planned to be added in the future
\ No newline at end of file
+  User/Account level apis are planned to be added in the future
```
Comment (tchaikov) [inline PendingReleaseNotes]: [off-topic] we need to fix the document to use APIs instead of apis to be consistent in another PR, as ReleaseNotes is use facing document.

Comment (ideepika) [reply PendingReleaseNotes]: on it!

Comment (tchaikov) [reply PendingReleaseNotes]: why? i think **Pending**ReleaseNotes are note released yet. so no need to backport them, i guess.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2020-03-03_13:02:55-rados-wip-kefu-testing-2020-03-03-1622-distro-basic-smithi/

the failure were either 

- caused by cephadm 
- fixed by #33701

File: PendingReleaseNotes (hunk 15 above)
Comment (ideepika) [reply PendingReleaseNotes]: yeah, thanks for clarifying :+1:

**Comment by ideepika** [comment]
jenkins test make check please

**Comment by ideepika** [comment]
jenkins test make check

File: PendingReleaseNotes [hunk 16]
```diff
@@ -316,3 +316,8 @@
   If the zonegroup is part of a realm, the change must be committed with
   'radosgw-admin period update --commit' - otherwise the change will take
   effect after radosgws are restarted.
+
+* Monitors now have a config option ``mon_allow_pool_size_one`` to allow
+  setting pool size to 1, which is disabled by default. However, if enabled,
+  the user now have to pass ``--yes-i-really-mean-it`` flag to ``osd pool set size 1``
+  if they really are sure of configuring pool size 1.
```
Comment (neha-ojha) [inline PendingReleaseNotes]:
```suggestion
  users now have to pass the ``--yes-i-really-mean-it`` flag to ``osd pool set size 1``, 
  if they are really sure about configuring a pool with no replicas.
```

**Comment by gregsfortytwo** [comment]
Still looks good; 👍 on waiting until Octopus I guess.

File: PendingReleaseNotes (hunk 16 above)
Comment (ideepika) [reply PendingReleaseNotes]: thanks, looks better put that way :+1:
